src/manager/lotus.rs

```diff
@@ -434,11 +461,233 @@ impl<T: JsonRpcClient + Send + Sync> LotusSubnetManager<T> {
+            .collect::<Result<_>>()
+    }
+
+    async fn checkpoint_template(&self, epoch: ChainEpoch) -> Result<NativeBottomUpCheckpoint> {
+        let template = self
+            .lotus_client
+            .ipc_get_checkpoint_template(&self.gateway_addr, epoch)
+            .await
+            .map_err(|e| {
+                anyhow!(
+                    "error getting bottom-up checkpoint template for epoch:{epoch:} due to {e:}"
+                )
+            })?;
+        let mut checkpoint = BottomUpCheckpoint::new(template.source().clone(), epoch);
+        checkpoint.data.children = template.data.children;
+        checkpoint.data.cross_msgs = template.data.cross_msgs;
+
+        NativeBottomUpCheckpoint::try_from(checkpoint)
+    }
+
+    async fn populate_prev_hash(
+        &self,
+        template: &mut NativeBottomUpCheckpoint,
+        subnet: &SubnetID,
+        previous_epoch: ChainEpoch,
+    ) -> Result<()> {
+        let response = self
+            .lotus_client
+            .ipc_get_prev_checkpoint_for_child(&self.gateway_addr, subnet)
+            .await
+            .map_err(|e| {
+                anyhow!(
+                    "error getting previous bottom-up checkpoint for epoch:{:} in subnet: {:?} due to {e:}",
+                    previous_epoch,
+                    subnet
+                )
+            })?;
+
+        // if previous checkpoint is set
+        if let Some(cid_map) = response {
+            let cid = Cid::try_from(cid_map)?;
+            template.prev_check = Some(cid.to_bytes());
+        } else {
+            template.prev_check = None;
+        }
+
+        Ok(())
+    }
+
+    async fn populate_proof(&self, template: &mut NativeBottomUpCheckpoint) -> Result<()> {
+        let proof = create_proof(&self.lotus_client, template.epoch).await?;
+        let bytes = cbor::serialize(&proof, "fvm bottom up checkpoint proof")?.to_vec();
+        template.proof = Some(bytes);
+        Ok(())
+    }
+
+    async fn submit(
+        &self,
+        validator: &Address,
+        checkpoint: NativeBottomUpCheckpoint,
+    ) -> Result<ChainEpoch> {
+        let to = checkpoint.source.subnet_actor();
+        let message = MpoolPushMessage::new(
+            to,
+            *validator,
+            ipc_subnet_actor::Method::SubmitCheckpoint as MethodNum,
+            cbor::serialize(&BottomUpCheckpoint::try_from(&checkpoint)?, "checkpoint")?.to_vec(),
+        );
+        let message_cid = self.lotus_client.mpool_push(message).await.map_err(|e| {
+            anyhow!(
+                "error submitting checkpoint for epoch {:} in subnet: {:?} with reason {e:}",
+                checkpoint.epoch,
+                checkpoint.source
+            )
+        })?;
+        log::debug!("checkpoint message published with cid: {message_cid:?}");
+
+        Ok(self.lotus_client.state_wait_msg(message_cid).await?.height as ChainEpoch)
+    }
+}
+
+mod conversion {
```

Comment:
Should we take this module into `fvm/conversion.rs` like we did with `evm/conversion.rs` or maybe add a `conversion` subfolder that has `conversion/fvm.rs` and `conversion/evm.rs`?

Reply:
yep, this works too. Initially I thought it's just a small mod, so just placed in place. Separating into another file works fine.